shared/lsp/registration.ts

```diff
@@ -363,203 +361,106 @@ export async function register({
-                ),
+                workspaceFolders: null,
                 initializationOptions,
             },
-            registerProviders: true,
+            registerProviders: false,
         })
         subscriptions.add(connection)
-        withConnection = async (workspaceFolder, fn) => {
-            let tempWorkspaceFolder: lsp.WorkspaceFolder | undefined
-            // If workspace folder is not known yet, add it
-            if (
-                !sourcegraph.workspace.roots.some(
-                    root => root.uri.toString() === workspaceFolder.href
-                )
-            ) {
-                tempWorkspaceFolder = {
-                    uri: workspaceFolder.href,
-                    name: path.posix.basename(workspaceFolder.pathname),
-                }
-                connection.sendNotification(
-                    lsp.DidChangeWorkspaceFoldersNotification.type,
-                    {
-                        event: {
-                            added: [tempWorkspaceFolder],
-                            removed: [],
-                        },
-                    }
-                )
-            }
-            try {
-                return await fn(connection)
-            } finally {
-                // If workspace folder was added, remove it
-                if (tempWorkspaceFolder) {
-                    connection.sendNotification(
-                        lsp.DidChangeWorkspaceFoldersNotification.type,
-                        {
-                            event: {
-                                added: [],
-                                removed: [tempWorkspaceFolder],
-                            },
-                        }
+        try {
+            return await fn(connection)
+        } finally {
+            connectionsByRootUri.delete(workspaceFolder.href)
+            connection.unsubscribe()
+        }
+    }
+    function addRoots(added: ReadonlyArray<sourcegraph.WorkspaceRoot>): void {
+        for (const root of added) {
+            const connectionPromise = (async () => {
+                try {
+                    const serverRootUri = clientToServerURI(
+                        new URL(root.uri.toString())
                     )
+                    const connection = await connect({
+                        clientRootUri: new URL(root.uri.toString()),
+                        initParams: {
+                            processId: null,
+                            rootUri: serverRootUri.href,
+                            capabilities: clientCapabilities,
+                            workspaceFolders: null,
+                            initializationOptions,
+                        },
+                        registerProviders: true,
+                    })
+                    subscriptions.add(connection)
+                    return connection
+                } catch (err) {
+                    logger.error('Error creating connection', err)
```

Comment:
Can we also log the root URI, so we can tell what the workspace is that failed?

Reply:
This is also unrelated to the PR. We can do this but in a separate effort.